app/Sources/TuistOnboarding/LogInViewModel.swift

```diff
@@ -0,0 +1,213 @@
+                    .queryItems?
+                    .first(where: { $0.name == "code" })?.value
+                else {
+                    continuation.resume(throwing: LogInViewModelError.missingAuthorizationCode)
+                    return
+                }
+
+                continuation.resume(returning: code)
+            }
+            authSession.presentationContextProvider = presentationContextProvider
+            authSession.prefersEphemeralWebBrowserSession = true
+            Task {
+                await MainActor.run {
+                    authSession.start()
+                }
+            }
+        }
+        if let code {
+            try await exchangeCodeForToken(code, codeVerifier: codeVerifier)
+        }
+    }
+
+    private func exchangeCodeForToken(
+        _ code: String,
+        codeVerifier: String
+    ) async throws {
+        let url = serverEnvironmentService.url().appending(
+            path: "oauth/token"
+        )
+
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+
+        let parameters = [
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": redirectURI,
+            "client_id": serverEnvironmentService.oauthClientId(),
+            "code_verifier": codeVerifier,
+        ]
+
+        let body =
+            parameters
+                .map {
+                    "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")"
+                }
+                .joined(separator: "&")
+
+        request.httpBody = body.data(using: .utf8)
+
+        let (data, response) = try await URLSession.shared.data(for: request)
+
+        guard let httpResponse = response as? HTTPURLResponse else {
+            throw LogInViewModelError.invalidTokenResponse
+        }
+
+        if httpResponse.statusCode == 200 {
+            try await handleTokenResponse(data)
+        } else {
+            throw LogInViewModelError.tokenExchangeFailed(statusCode: httpResponse.statusCode)
+        }
+    }
+
+    private func handleTokenResponse(_ data: Data) async throws {
+        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
+            throw LogInViewModelError.invalidTokenResponse
+        }
+
+        guard let accessToken = json["access_token"] as? String,
+              let refreshToken = json["refresh_token"] as? String
+        else {
+            throw LogInViewModelError.missingTokens
+        }
+
+        try await serverCredentialsStore.store(
+            credentials: ServerCredentials(
+                token: nil,
+                accessToken: accessToken,
+                refreshToken: refreshToken
+            ),
+            serverURL: serverEnvironmentService.url()
+        )
```

Comment:
Are both the CLI and the app sharing the same storage? I think for apps makes more sense to use the keychain scoped under the app domain, since there people expect models from the system asking to give keychain permission to the app.